### PR TITLE
Substitute native keccak with Merkle commitment on data integrity hash

### PR DESCRIPTION
**Pull request by LiuJiazheng**
Motivation:
The DecentralizedKV is divided by term MaxKvValue. User uses put semantics to put a trunk data in someone's disk, whose size is exactly MaxKvValue. Nowadays, the value is 4K, for current SSD minimal fetching unit is 4KB. However, Danksharding proposal supports a 128KB per BLOB.

For first time put, we cannot aviod to upload full data to generate a hash commiment. But we can reduce gas on submitting RoRA (via Dagger-Hashimoto) proof, by changing a full-data hash commitment to somewhat inclusion proof, and yes, Merkle tree can be an ideal tool.

By this changing, each time miner being asked to submit PoRA, it can just choose 4KB pieces of whole data, and then generate Merkle proof, thus submitting proofs, id it chooses and 4KB piece of data. On chain, KV contract does an Merkle inclusion verification. The effect is identical as previous one, in term of demand of Proof of Random Access.

Change:
KV contract, extends bytes24 hash to a bytes32 standard hash after keccak256 TestSystemContract, change both verify functions
Add tests about submitting proofs + verifying

### CODE REVIEW

**Comment by LiuJiazheng** [comment]
It is not an end but I think it is a good time to split PR. And 
NOTICE: It dpes not COMPLETELY solve problem------ in fact it even introduces an attack dimension.

Now we allow miner to submit PoRA partially, not as wholesale. Then there emerges a new lazy validator issue, even in Dagger-Hashimoto version. 
Suppose miner gets a hash0 and wants to submit a new turn of PoRA. Let's also assume our data trunkdefined in DecentrailizedKV contract is 128KB. 
Now there is a lazy guy, instead of storing full 128KB data, he only stores first 4KB data per data trunk, and following 31 pieces are in form of Merkle proofs. Now when he submits PoRA, he can still choose a series like 12-4-7-9 (suppose we only need 4 random check points), and only submits first 4KB, and proofs and idx is always zero. So he can purge all 128-4 = 124KB data and still cheat on system, which is a case we definitely don't want to see.

The reason is, we relax too much on which part of 4KB data miners shall submit inside a data trunk, let's say, 128KB. We have to take this kind of decision making freedom back, more precisely, our Hashimoto-ish algorithm shall reflect this randomness, not in shard-wise, but inside a trunk, there shall be another random index selection based on hash it generates.

File: contracts/TestDaggerHashPrecompile.sol [hunk 1]
```diff
@@ -3,6 +3,7 @@ pragma solidity ^0.8.0;
 
 import "./TestDecentralizedKV.sol";
 import "./CallerLibraries.sol";
+import "hardhat/console.sol";
```
Comment (LiuJiazheng) [inline contracts/TestDaggerHashPrecompile.sol]: need to be removed

File: contracts/DKVDaggerHashimoto.sol [hunk 2]
```diff
@@ -212,7 +251,9 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
             uint256 parent = uint256(hash0) % rows;
             uint256 kvIdx = parent + (startShardId << shardEntryBits);
             bytes memory data = maskedData[i];
-            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data), "invalid access proof");
+            require(systemContract.checkDaggerDataWithProof(checkIdList[i],
```
Comment (LiuJiazheng) [inline contracts/DKVDaggerHashimoto.sol]: Here, the idx inside a Merlke tree node is specificed by user, hence introducing a lazy miner attack. See my second comment of this pr.

what if we change this checkIdList[i] to randomIdxMerkleNode, where
randomIdxMerkleNode = kvIdx % (maxKVSize >> 4K).
Is this helping to preventing lazy validator attack and does not break general randomness in Hashimoto algorithm?

File: contracts/TestDaggerHashPrecompile.sol (hunk 1 above)
Comment (LiuJiazheng) [inline contracts/TestDaggerHashPrecompile.sol]: to be removed

File: contracts/TestDaggerHashPrecompile.sol [hunk 3]
```diff
@@ -49,6 +50,9 @@ contract TestKVWithDaggerHash is TestDecentralizedKV {
         assembly {
             paddrValue := sload(paddr.slot)
         }
+        // This optimization is only working for we have a PhyAddr with byte32 or less length
+        // To use Merkle tree commitment, we extend a hash field from Bytes24 to Bytes32
+        // This one immediately goes invalid
```
Comment (LiuJiazheng) [inline contracts/TestDaggerHashPrecompile.sol]: @qizhou Is this comment stating right? Now a PhyAddr is larger than Bytes32, the gas-saving optimization seems not valid anymore?

File: contracts/DKVDaggerHashimoto.sol [hunk 4]
```diff
@@ -195,15 +196,53 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+        uint256 startShardId,
+        uint256 shardLenBits,
+        bytes32 hash0,
+        uint256[] memory checkIdList,
+        bytes32[][] memory proofsDim2,
+        bytes[] memory maskedData
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        require(checkIdList.length == randomChecks, "proofs idx vs checks: length mismatch");
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: I don't think we should pass checkIdList - instead, `rows` will be the number of 4KB chunks in a shard, and we will find `chunkIdx` instead of `kvIdx`, and we will also know the position of the chunk in the large value of the KV.

**Comment by qizhou** [comment]
> It is not an end but I think it is a good time to split PR. And NOTICE: It dpes not COMPLETELY solve problem------ in fact it even introduces an attack dimension.
> 
> Now we allow miner to submit PoRA partially, not as wholesale. Then there emerges a new lazy validator issue, even in Dagger-Hashimoto version. Suppose miner gets a hash0 and wants to submit a new turn of PoRA. Let's also assume our data trunkdefined in DecentrailizedKV contract is 128KB. Now there is a lazy guy, instead of storing full 128KB data, he only stores first 4KB data per data trunk, and following 31 pieces are in form of Merkle proofs. Now when he submits PoRA, he can still choose a series like 12-4-7-9 (suppose we only need 4 random check points), and only submits first 4KB, and proofs and idx is always zero. So he can purge all 128-4 = 124KB data and still cheat on system, which is a case we definitely don't want to see.
> 
> The reason is, we relax too much on which part of 4KB data miners shall submit inside a data trunk, let's say, 128KB. We have to take this kind of decision making freedom back, more precisely, our Hashimoto-ish algorithm shall reflect this randomness, not in shard-wise, but inside a trunk, there shall be another random index selection based on hash it generates.

The key is that PoRA will run on chunk index instead of KV index - that said it will already run on 4KB random access, and then we could derive which KV index it is accessing from the chunk index.  This should address the concern here.

File: contracts/DecentralizedKV.sol [hunk 5]
```diff
@@ -104,16 +104,11 @@ contract DecentralizedKV {
...
         kvMap[skey] = paddr;
 
-        // Weird that cannot call precompiled contract like this (solidity issue?)
-        // storageManager.putRaw(paddr.kvIdx, data);
-        // Use call directly instead.
-        (bool success, ) = address(storageManager).call(
-            abi.encodeWithSelector(IStorageManager.putRaw.selector, paddr.kvIdx, data)
-        );
-        require(success, "failed to putRaw");
+        storageManager.putRaw(paddr.kvIdx, data);
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: I would recommend keeping the original code as we found an issue when running on the actual network.

File: contracts/DecentralizedKV.sol [hunk 6]
```diff
@@ -11,6 +13,7 @@ contract DecentralizedKV {
     uint256 public immutable startTime;
     uint256 public immutable maxKvSize;
     uint40 public lastKvIdx = 0; // number of entries in the store
+    uint32 public constant SYSTEM_BLOCK_SIZE = 4096; // 4K bytes is normal SSD minimal fetchable size
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: `SYSTEM_BLOCK_SIZE` => `CHUNK_SIZE` for consistency with technical white paper.

File: contracts/DecentralizedKV.sol [hunk 7]
```diff
@@ -20,7 +23,7 @@ contract DecentralizedKV {
         /* Block Size. aligned with 2^n */
         uint24 kvSize;
         /* Commitment */
-        bytes24 hash;
+        bytes32 hash;
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: I think we should still use bytes24 as a truncated root hash could still work for inclusion proof.

**Comment by LiuJiazheng** [comment]
Second comprehensive commit updated.

Major focus : function `_generateChunkIdx` helps generating a chunkIdx within a KV block, as offset whose unit is 4KB. @qizhou : Does it look good to you?

For some reasons we cannot remove checkIdList in method `mine` from DKVDaggerHashimoto.sol, otherwises we will get a  "Stack too deep" error. Personally I believe it is more due to a compilation issue rather our implementation issue. Just leave it as it is.

File: contracts/TestDaggerHashPrecompile.sol [hunk 8]
```diff
@@ -86,4 +88,15 @@ contract TestKVWithDaggerHash is TestDecentralizedKV {
         PhyAddr memory paddr = kvMap[skey];
         return sysContract.checkDaggerData(paddr.kvIdx, paddr.hash, maskedData);
     }
+
+    function checkDaggerHashNormalMerkle(uint256 idx,
+                                         bytes32 key,
+                                         bytes32[] memory proofs,
+                                         bytes memory maskedData 
+    )public view returns (bool) {
```
Comment (qizhou) [inline contracts/TestDaggerHashPrecompile.sol]:
```suggestion
    ) public view returns (bool) {
```

File: contracts/DKVDaggerHashimoto.sol [hunk 9]
```diff
@@ -195,15 +196,18 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
     function _hashimotoKeccak256(
         uint256 startShardId,
         uint256 shardLenBits,
         bytes32 hash0,
         bytes[] memory maskedData
     ) internal view returns (bytes32) {
-        require(maskedData.length == randomChecks, "incorrect PoRA");
+        /* Assumption check */
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        if (randomChecks != 0) require(maskedData[0].length <= CHUNK_SIZE, "too large data uploaded");
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: I guess we should not allow` randomCheck == 0`?

File: contracts/DecentralizedKV.sol [hunk 10]
```diff
@@ -104,7 +102,8 @@ contract DecentralizedKV {
             lastKvIdx = lastKvIdx + 1;
         }
         paddr.kvSize = uint24(data.length);
-        paddr.hash = bytes24(keccak256(data));
+        uint256 nChunkBits = generateChunkBits(data.length);
+        paddr.hash = bytes24(MerkleLib.merkleRoot(data, CHUNK_SIZE, nChunkBits));
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: Could we have an optimized `merkleRootWithMinTree()` so that it just calculates the root of the data with minimal tree size (instead of doing extra calculations?)

File: contracts/DKVDaggerHashimoto.sol [hunk 11]
```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+        uint256 rows = 1 << (shardEntryBits + shardLenBits);
+        uint256 parent = uint256(hash0) % rows;
+        uint256 kvIdx = parent + (startShardId << shardEntryBits);
+        return kvIdx;
+    }
+
+    function _generateChunkIdx(
+        bytes32 hash0
+    ) internal view returns (uint256) {
+        require((CHUNK_SIZE != 0) && ((CHUNK_SIZE & (CHUNK_SIZE - 1)) == 0), "CHUNK_SIZE invalid");
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: We should not check 2**n of CHUNK_SIZE here - it should be in the ctor code so that we don't need to pay the gas per call.

File: contracts/DKVDaggerHashimoto.sol [hunk 12]
```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 chunksPerKVBlock = maxKvSize / CHUNK_SIZE;
+        /* If maxKvSize is smaller than CHUNK_SIZE, early exit */
+        if (chunksPerKVBlock == 0) return 0;
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: This should never happen - or we could put it as immutable so that we don't need to evaluate the number per call.

File: contracts/DKVDaggerHashimoto.sol [hunk 13]
```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        require(maskedData.length <= CHUNK_SIZE, "calldata too large");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+
+        for (uint256 i = 0; i < randomChecks; i++) {
+            require(maskedData[i].length == maxKvSize, "invalid proof size");
+            uint256 kvIdx = _generateKVIdx(hash0, startShardId, shardLenBits);
+            uint256 chunkIdx = _generateChunkIdx(hash0);
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: I guess this may be wrong - a kvIdx = 12 will always access chunk 12 (suppose max value size is 128KB), where we want the access to be uniform.

Instead, we should have

`
uint256 rows = 1 << (shardEntryBits + shardLenBits + chunkLenBits);
uint256 parent = uint256(hash0) % rows;
uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
uint256 kvIdx = chunkIdx >> chunkLenBits;

**Review by qizhou — COMMENTED** [review]
You can use `npx prettier --write` to format the code.

File: contracts/DKVDaggerHashimoto.sol (hunk 9 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: It is a setting from config. Cannot assume that, I believe

File: contracts/DKVDaggerHashimoto.sol (hunk 11 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: Make senses.

File: contracts/DKVDaggerHashimoto.sol (hunk 12 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: In fact, it can happen, at least in our test env. 
In test env we have maxKvSize around 64B - 128B, smaller than default CHUNK_SIZE 4k.
Under above situation, Hashimoto with Merkle proof degenerates one simple keccak256. 
But the corner case remains.

File: contracts/DKVDaggerHashimoto.sol (hunk 9 above)
Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol]: It is a constant (immutable) of the contract, where the non-zero check should be also in ctor.

File: contracts/DKVDaggerHashimoto.sol (hunk 12 above)
Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol]: Should not in that case that CHUNK_SIZE = 64B or 128B?

File: contracts/DecentralizedKV.sol (hunk 10 above)
Comment (LiuJiazheng) [reply contracts/DecentralizedKV.sol]: Could you please elaborate?
Shouldn't a standard Merkle tree spanned by a length of data.length / CHUNK_SIZE ?

File: contracts/DKVDaggerHashimoto.sol (hunk 13 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: We haven't changed too much. The caculation method of kvIdx persists:
```
function _generateKVIdx(
        bytes32 hash0,
        uint256 startShardId,
        uint256 shardLenBits
    ) internal view returns (uint256) {
        uint256 rows = 1 << (shardEntryBits + shardLenBits);
        uint256 parent = uint256(hash0) % rows;
        uint256 kvIdx = parent + (startShardId << shardEntryBits);
        return kvIdx;
    }
```
It is uniformally distributed as before. 
The tricky part is chunkIdx. Here is story of what I understand:
Now we have hierarchy : shard - KV block - chunks
A shard has multiple KV blocks (say 128KB per block), a block has 32 chunks (say 4K per chunk). 
If we count everything in chunk, yes, term `rows` represents how many chunks we focus and use hash0 to get an offset (parent), then a base position added --- we are done. 

However if we examine the API of MerkleLib.getProof, by design, it accepts a chunkIdx **within** the a KV Block, where in our example above, it should be range from 0 to 31. 

So if we caculate:
uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
it still needs:
```
chunkIdx = chunkIdx % (1 <<chunkLenBits ) 
               = (parent + (startShardId << (shardEntryBits + chunkLenBits)) ) % ( 1 <<chunkLenBits )
               = parent % (1 << chunkLenBits)
               = uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) % (1 << chunkLenBits)
               = uint256(hash0) % (1 << chunkLenBits)
```
since (1 << chunkLenBits) is a factor of (1 << (shardEntryBits + shardLenBits + chunkLenBits));
That's my function _generateChunkIdx.

```
uint256 kvIdx = chunkIdx >> chunkLenBits
                      =  ( parent + (startShardId << (shardEntryBits + chunkLenBits)) ) >> chunkLenBits
                      =  ( uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) >> chunkLenBits ) +      
                          (startShardId << shardEntryBits)
```
 where uint256(hash0) % (1 << (shardEntryBits + shardLenBits)) =/=  ( uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) >> chunkLenBits, can't quite tell which one is more evenly distributed. 

My naive thinking is their Lebesgue measure should be same.............

Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: At least can we make sure chunkIdx calculation is correct..........

**Comment by LiuJiazheng** [comment]
Uploaded a test case. Still having some discussions with Qi. Hold on.

File: contracts/DKVDaggerHashimoto.sol (hunk 9 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: Oh you mean we can get it at ctor. Surely, it could be better, saving gas.

File: contracts/DKVDaggerHashimoto.sol (hunk 12 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: Currently we make CHUNK_SIZE a constant but yep I guess that's basically a style issue. Perhaps we can change that.

File: contracts/DecentralizedKV.sol (hunk 10 above)
Comment (qizhou) [reply contracts/DecentralizedKV.sol]: `merkleRoot()` currently asked for the size of the tree as the tree can be arbitrarily bigger as long as its leafs cover all the data leaves.

E.g., a data with 5 leaves will need a tree with minimal of 8 leaves, but the trees will 16, 32, 64 leaves also work.

File: contracts/DKVDaggerHashimoto.sol (hunk 12 above)
Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol]: Yes, we should also stick to the assumption that max value size is a 2^n multiple of CHUNK_SIZE.  Further, we should minimize the checks in the contract call methods that can be called frequently.

File: contracts/DKVDaggerHashimoto.sol (hunk 13 above)
Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol]: What I mean here is not `P(chunkIdx)`, where chunkIdx is uniform-distributed in [0,31], I mean `P(chunkIdx | kvIdx)` is not uniform-distributed, e.g., if kvIdx = 13, then `P(chunkIdx = 13 | kvIdx = 13) = 1`, while `P(chunkIdx \neq 13 | kvIdx = 13) = 0`.

The key here is the lowest bits are reused in both chunkIdx and kvIdx, which does not provide sufficient entropy as chunkIdx and kvIdx correlated.

Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: That makes sense!
I didn't consider conditional possibility. You are right.

File: contracts/DecentralizedKV.sol (hunk 10 above)
Comment (LiuJiazheng) [reply contracts/DecentralizedKV.sol]: So the `nChunkBits` control how big a tree can be, right?
Right now, the calculation is, for example,
Data.length divided by CHUNK_SIZE, gets a chunk number, say 5, we will round up next number which is of 2^n, 8, then the bitslen is 3. 
As long as we control `nChunkBits`, we control the Merkle tree spanning.

Comment (qizhou) [reply contracts/DecentralizedKV.sol]: Yes, but what I mean is that we could optimize the calculation in `merkleRootWithMinTree()` which returns the root hash of the minimal tree without `nChunkBits` as a parameter.

Comment (LiuJiazheng) [reply contracts/DecentralizedKV.sol]: Okay I had an offline talk with Qi. Now I know the request is to simplify the API........
It should be quite intuitive.

File: contracts/DKVDaggerHashimoto.sol [hunk 14]
```diff
@@ -195,15 +201,18 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
     function _hashimotoKeccak256(
         uint256 startShardId,
         uint256 shardLenBits,
         bytes32 hash0,
         bytes[] memory maskedData
     ) internal view returns (bytes32) {
-        require(maskedData.length == randomChecks, "incorrect PoRA");
+        /* Assumption check */
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(maskedData[0].length <= chunkSize, "too large data uploaded");
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: maskedData length should equal to chunkSize so that the masked data has full entropy.

File: contracts/DKVDaggerHashimoto.sol [hunk 15]
```diff
@@ -43,11 +44,16 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+        /* Assumptions */
+        require(_config.shardSizeBits >= _config.maxKvSizeBits, "shardSize too small");
+        require(_config.randomChecks > 0, "At least one checkpoint needed");
+
         systemContract = _config.systemContract;
         shardSizeBits = _config.shardSizeBits;
         maxKvSizeBits = _config.maxKvSizeBits;
-        require(_config.shardSizeBits >= _config.maxKvSizeBits, "shardSize too small");
         shardEntryBits = _config.shardSizeBits - _config.maxKvSizeBits;
+        chunkLenBits = _config.maxKvSizeBits < DEFAULT_CHUNK_SIZE_BITS ?
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: I think it may be better to move chunkLenBits to DecentralizedKV.sol and replace DEFAULT_CHUNK_SIZE_BITS?  It sounds to me that 0 chunkLenBits is invalid.

File: contracts/DKVDaggerHashimoto.sol [hunk 16]
```diff
@@ -223,6 +232,45 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+        bytes32[][] memory proofsDim2,
+        bytes[] memory maskedData
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 rows = 1 << (shardEntryBits + shardLenBits + chunkLenBits);
+
+        for (uint256 i = 0; i < randomChecks; i++) {
+            require(maskedData[i].length <= chunkSize, "invalid proof size");
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: See above.

File: contracts/MerkleLib.sol [hunk 17]
```diff
@@ -55,7 +55,13 @@ library MerkleLib {
...
             chunkIdx = chunkIdx / 2;
         }
-        return hash == root;
+        /* NOTICE: Due to our design of PhyAddr, only front 24 bytes
+         *         are valid. We only validate that part.
+         * With no doubt, it introduces some vulnerability compared 
+         * with a standard Merkle validation. It is a trade-off,
+         * if we want to put all meta data into a single bytes32(opcode length)
+         */
+        return bytes24(hash) == bytes24(root);
```
Comment (qizhou) [inline contracts/MerkleLib.sol]: Could we add a method like `calculateRootWithProof`, which returns the root so that the comparison logic can be done by application contract?  MerkleLib should not be designed to understand the size of hash.

File: contracts/DKVDaggerHashimoto.sol [hunk 18]
```diff
@@ -223,6 +232,45 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
+            uint256 parent = uint256(hash0) % rows;
+            uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
+            uint256 kvIdx = chunkIdx >> chunkLenBits;
+            chunkIdx = chunkIdx % (1 << chunkLenBits);
+            bytes memory data = maskedData[i];
+            /* NOTICE: Now we have kvIdx and chunkIdx both generated from hash0
+             *         The difficulty should increase intrinsically */
+            require(
+                systemContract.checkDaggerDataWithProof(chunkIdx, kvMap[idxMap[kvIdx]].hash, proofsDim2[i], data),
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol:263]: We are changing the API to `systemContract.checkDaggerData` to `systemContract.unmaskData`, which will return unmasked data so that the inclusion verification check can be done in this contract.

PR is here https://github.com/QuarkChain/go-ethereum/pull/110

**Review by qizhou — COMMENTED** [review]
A lot of good changes.  Almost there 👍

File: contracts/DKVDaggerHashimoto.sol (hunk 15 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: It should keep aligned with _config.maxKvSizeButs, since there clearly exists a dependency in between. Unless we move maxKvSizeBits into DecentralizedKV.sol, otherwise it does not make too much sense.
I'd rather keep it here.

File: contracts/DKVDaggerHashimoto.sol (hunk 18 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol:263]: Can we make it as a separate pr? Considering it not quite related to this topic, I believe it is an orthogonal effort.
We might change all involved functions call and tests.

Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol:263]: Yes, just FYI.

File: contracts/DecentralizedKV.sol [hunk 19]
```diff
@@ -40,19 +43,16 @@ contract DecentralizedKV {
...
-                v = (v * fp) >> 128;
-            }
-            fp = (fp * fp) >> 128;
-            n = n / 2;
-        }
-        return v;
+        return BinaryRelated.pow(fp, n);
+    }
+
+    function generateChunkBits(uint256 dataLen) internal view returns (uint256) {
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: This method is not used?

File: contracts/MerkleLib.sol [hunk 20]
```diff
@@ -37,6 +39,41 @@ library MerkleLib {
         return nodes[0];
     }
 
+    function merkleRootWithMinTree(
+        bytes memory data,
+        uint256 chunkSize
+    ) internal pure returns (bytes32) {
+        uint256 n = (data.length + chunkSize - 1) / chunkSize;
+        /* Round up to next 2^n */
+        uint256 nChunks = (n <= 1) ? 1 : BinaryRelated.findNextPowerOf2(n);
```
Comment (qizhou) [inline contracts/MerkleLib.sol]: We don't need to find the next power of 2 - just allocate nodes with n, and when then merge two nodes (in line 68-70), if the last node is odd, then merge it with bytes32(0).

File: contracts/TestDaggerHashPrecompile.sol (hunk 8 above)
Comment (qizhou) [reply contracts/TestDaggerHashPrecompile.sol]: Looks it is not yet solved?

File: contracts/DKVDaggerHashimoto.sol [hunk 21]
```diff
@@ -322,14 +371,18 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
...
     // We allow cross mine multiple shards by aggregating their difficulties.
+    // For some reasons, we never use checkIdList but if we remove it, we will get
+    // a `Stack too deap error`
     function mine(
         uint256 startShardId,
         uint256 shardLenBits,
         address miner,
         uint256 minedTs,
         uint256 nonce,
+        uint256[] memory checkIdList,
```
Comment (qizhou) [inline contracts/DKVDaggerHashimoto.sol]: It seems that the overflow error can be fixed by upgrading solidity to 0.8.16 (the latest hardhat supported version)

File: contracts/DKVDaggerHashimoto.sol (hunk 15 above)
Comment (qizhou) [reply contracts/DKVDaggerHashimoto.sol]: Oh, what I mean is to replace DEFAULT_CHUNK_SIZE_BITS with an immutable chunkSizeBits, and put require(maxKvSizeBits >= chunkSizeBIts, "...") in the ctor.

File: contracts/DecentralizedKV.sol (hunk 19 above)
Comment (LiuJiazheng) [reply contracts/DecentralizedKV.sol]: Yes, will remove it

File: contracts/TestDaggerHashPrecompile.sol (hunk 8 above)
Comment (LiuJiazheng) [reply contracts/TestDaggerHashPrecompile.sol]: Somehow missed applying linter here. Appiled just now.

File: contracts/DKVDaggerHashimoto.sol (hunk 21 above)
Comment (LiuJiazheng) [reply contracts/DKVDaggerHashimoto.sol]: Confirmed.

File: contracts/MerkleLib.sol (hunk 20 above)
Comment (LiuJiazheng) [reply contracts/MerkleLib.sol]: Okay. But data.length == 0 is still a corner case.

Comment (qizhou) [reply contracts/MerkleLib.sol]: Yep.  Just return bytes32(0) when data.length == 0 is a good practice.

File: contracts/MerkleLib.sol [hunk 22]
```diff
@@ -37,6 +37,40 @@ library MerkleLib {
...
+                len = chunkSize;
+            }
+            assembly {
+                hash := keccak256(add(add(data, 0x20), off), len)
+            }
+            nodes[i] = hash;
+        }
+        uint256 n = nChunks;
+        while (n != 1) {
+            for (uint256 i = 0; i < n / 2; i++) {
```
Comment (qizhou) [inline contracts/MerkleLib.sol:68]: I am not sure this would work if n is odd?

File: contracts/BinaryRelated.sol [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+    function getExponentiation(uint256 n) internal pure returns (uint256) {
+        uint256 cnt = 0;
+        while(n != 0) {
+            n = n >> 1;
+            cnt++;
+        }
+        return cnt-1;
+    }
+
+    function getBitsLen(uint256 n) internal pure returns (uint256) {
```
Comment (qizhou) [inline contracts/BinaryRelated.sol]: This function can be removed?

File: contracts/BinaryRelated.sol [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+        return v;
+    }
+
+    function findNextPowerOf2(uint256 n) internal pure returns (uint256) {
+        n = n-1;
+        while ((n & (n-1) != 0)) n = n & (n-1);
+        return n << 1;
+    }
+
+    function getExponentiation(uint256 n) internal pure returns (uint256) {
```
Comment (qizhou) [inline contracts/BinaryRelated.sol]: This function can be removed?

File: contracts/BinaryRelated.sol [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+                v = (v * fp) >> 128;
+            }
+            fp = (fp * fp) >> 128;
+            n = n / 2;
+        }
+        return v;
+    }
+
+    function findNextPowerOf2(uint256 n) internal pure returns (uint256) {
+        n = n-1;
```
Comment (qizhou) [inline contracts/BinaryRelated.sol:19]: This function can be removed?

File: contracts/DecentralizedKV.sol [hunk 26]
```diff
@@ -10,14 +12,16 @@ contract DecentralizedKV {
     uint256 public immutable dcfFactor;
     uint256 public immutable startTime;
     uint256 public immutable maxKvSize;
+    uint256 public immutable chunkSize;
     uint40 public lastKvIdx = 0; // number of entries in the store
+    uint256 public constant DEFAULT_CHUNK_SIZE_BITS = 12; // 4K bytes is normal SSD minimal fetchable size
```
Comment (qizhou) [inline contracts/DecentralizedKV.sol]: Can be removed?

**Review by qizhou — COMMENTED** [review]
Great job!  Just a few comments.

File: contracts/MerkleLib.sol (hunk 22 above)
Comment (LiuJiazheng) [reply contracts/MerkleLib.sol:68]: I don't think it can. Let's resume it to 2^n, so that it is 100% safe.
`chunkSize = Binaryrelated.getExponentiation(BinaryRelated.findNextPowerOf2(n))`

**Comment by LiuJiazheng** [comment]
How was it? Can it be closed hereby?

File: contracts/BinaryRelated.sol [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (qizhou) [inline contracts/BinaryRelated.sol:19]: I think there is a more efficient algorithm here:

https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2

In fact, we may not need the function if merkleRootMinTree can use an adaptive way to merge two hashes.

**Review by qizhou — APPROVED** [review]
This is a comment regarding code optimization, but we can do it in the next iteration.
